### PR TITLE
FIX: App crashes when password is null or tags are retrieved from parcel FIXES #25

### LINKED ISSUE #25: Crashes at wakeup
Bunch of reports with this stacktrace. Might be happening because we don't persist password, and during wakeup its gone from memory.

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.twistedplane.sealnote/com.twistedplane.sealnote.NoteActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String[] java.lang.String.split(java.lang.String)' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3319)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3415)
	at android.app.ActivityThread.access$1100(ActivityThread.java:229)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1821)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:7325)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String[] java.lang.String.split(java.lang.String)' on a null object reference
	at com.twistedplane.sealnote.data.Note.convertToTagSet(Note.java:118)
	at com.twistedplane.sealnote.data.Note.readFromParcel(Note.java:97)
	at com.twistedplane.sealnote.data.Note.<init>(Note.java:76)
	at com.twistedplane.sealnote.data.Note$1.createFromParcel(Note.java:250)
	at com.twistedplane.sealnote.data.Note$1.createFromParcel(Note.java:248)
	at android.os.Parcel.readParcelable(Parcel.java:2367)
	at android.os.Parcel.readValue(Parcel.java:2264)
	at android.os.Parcel.readArrayMapInternal(Parcel.java:2614)
	at android.os.BaseBundle.unparcel(BaseBundle.java:221)
	at android.os.BaseBundle.getBoolean(BaseBundle.java:659)
	at android.app.Activity.restoreHasCurrentPermissionRequest(Activity.java:7153)
	at android.app.Activity.performCreate(Activity.java:6903)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1136)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3266)
	... 9 more
```

```
java.lang.RuntimeException: An error occured while executing doInBackground()
	at android.os.AsyncTask$3.done(AsyncTask.java:304)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:355)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:222)
	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.lang.IllegalArgumentException: Password null or not acceptable
	at com.twistedplane.sealnote.data.DatabaseHandler.update(DatabaseHandler.java:84)
	at com.twistedplane.sealnote.data.DatabaseHandler.getWritableDatabase(DatabaseHandler.java:110)
	at com.twistedplane.sealnote.data.DatabaseHandler.getReadableDatabase(DatabaseHandler.java:115)
	at com.twistedplane.sealnote.data.DatabaseHandler.getNotesCursor(DatabaseHandler.java:558)
	at com.twistedplane.sealnote.data.AdapterLoader.loadInBackground(AdapterLoader.java:30)
	at com.twistedplane.sealnote.data.AdapterLoader.loadInBackground(AdapterLoader.java:13)
	at android.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:312)
	at android.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:69)
	at android.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:57)
	at android.os.AsyncTask$2.call(AsyncTask.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	... 3 more
```

### PR DESCRIPTION
**Pull request by rohitramkumar308**
- Check for null password and navigate to PasswordActivity if true. When app permissions are revoked manually, the password in the memory is lost and the app crashes.

- Pass note to NoteContent.FromString. Tags are passed to NoteContent instead of note that cause the tags to be null when read from parcel.

### CODE REVIEW

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 1]
```diff
@@ -107,46 +107,51 @@ public void afterTextChanged(Editable editable) {
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_note);
-        // Even though we change content view later, we secure window as soon as possible
-        Misc.secureWindow(NoteActivity.this);
 
-        mBackgroundColor = 0;
-        mAutoSaveEnabled = PreferenceHandler.isAutosaveEnabled(NoteActivity.this);
+        if(SealnoteApplication.getDatabase().getPassword() == null){
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java]: (1) You can avoid the wrapping rest of the code inside else by writing code like this. This also avoids making the diff bigger.

```
if (cond) {
  stms ...
  return;
}

else-stms...
```

(2) Since this is used twice, maybe you can abstract it as a function like "isPasswordLoaded".or something like that

File: Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java [hunk 2]
```diff
@@ -78,36 +80,39 @@
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         Log.d(TAG, "Creating SealNote fragment...");
+        if(SealnoteApplication.getDatabase().getPassword()==null){
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java]: See (1)

File: Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java [hunk 3]
```diff
@@ -216,4 +221,9 @@ public void onLoaderReset(Loader<Cursor> cursorLoader) {
         mAdapter.clearCursor();
         layoutProgressHeader.setVisibility(View.VISIBLE);
     }
+    private void startPasswordActivity() {
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java]: Add a blank line before. Each function must be separated by a blank line.

File: Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java (hunk 2 above)
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java]: Spaces before and after "==" operator. Space after if.

File: Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java [hunk 4]
```diff
@@ -93,7 +93,7 @@ private void readFromParcel(Parcel inParcel) {
         mArchived = inParcel.readInt() > 0;
         mDeleted = inParcel.readInt() > 0;
         mType = Type.valueOf(inParcel.readString());
-        mNote = NoteContent.fromString(mType, inParcel.readString());
+        mNote = NoteContent.fromString(mType, note);
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java]: Why did you remove the `inParcel.readString()` and replaces with note? `note` is used just once here, so original code seems fine.

**Review by vishesh — CHANGES_REQUESTED** [review]
As for the code styling, you can look at the code in the file you are editing and get an idea. In general, google's style guide is fine[1].Just use 4 spaces for indentation, and horizontal alignment is recommended when it makes sense but not enforced.

@akshatgoel will be putting up a style guide soon.

Also, for the changes you don't have to close previous pull request and open new one. In general evreything is editable. As for the commit itself,  you can force push  (`git push --force`, be careful doing this) to your branch and pull request will update itself with your latest branch.

[1] https://google.github.io/styleguide/javaguide.html

File: Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java (hunk 4 above)
Comment (rohitramkumar308) [reply Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java]: The number of `writeString` and `readString` don't match up in the original code. `note` variable is never used. `NoteContent` must have the actual note with the type right ?

Comment (vishesh) [reply Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java]: Oh, I didn't see that in the original code had a note variable. Since that variable is used just once otherwise, just remove that dead binding, and keep the `mNote = NoteContent.fromString(mType, inParcel.readString());` line.

File: Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java [hunk 5]
```diff
@@ -107,6 +107,12 @@ public void afterTextChanged(Editable editable) {
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_note);
+
+        if(!Misc.isPasswordLoaded()){
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/NoteActivity.java:111]: (1) Please add a whitespace between "if" and the parenthesis.

https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

File: Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java [hunk 6]
```diff
@@ -79,6 +80,11 @@ public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         Log.d(TAG, "Creating SealNote fragment...");
 
+        if(!Misc.isPasswordLoaded()){
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/fragment/SealnoteFragment.java:83]: Refer (1)

File: Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java [hunk 7]
```diff
@@ -102,12 +101,12 @@ public void writeToParcel(Parcel outParcel, int flags) {
         outParcel.writeInt(mId);
         outParcel.writeInt(mPosition);
         outParcel.writeString(mNoteTitle);
+        outParcel.writeString(mType.name());
```
Comment (vishesh) [inline Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java:104]: Any reason this was moved from line 110 to this line? If not its better to keep code where it was to make diffs small.

**Review by vishesh — COMMENTED** [review]
Excellent. Only couple more small style issues. Lastly, could you please squash the commits together into single commit as they both address same issue. If you are not familiar with git squashing try these links -
 with git squashing try these links -

- https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
- http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

File: Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java (hunk 7 above)
Comment (rohitramkumar308) [reply Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java:104]: This is to remove that `note` variable and pass it directly into `fromString`

Comment (vishesh) [reply Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java:104]: Oh alright. You are maintaining the order of read/write in parcel. Looks good then.

Comment (rohitramkumar308) [reply Sealnote/src/main/java/com/twistedplane/sealnote/data/Note.java:104]: Created a new pull request with just one commit.
